### DISCUSSION TITLE
How to detect text on paste, and replace with a node?

### DISCUSSION THREAD

**Discussion by rfgamaral** [ueberdosis/tiptap#2053] [General Questions]
I need to detect certain text keywords and convert them into nodes, and that can be easily achievable with `nodeInputRule`. However, what if I just paste some text which contains the text keywords I need to convert into nodes? There's no such thing as `nodePasteRule`, so how can this be achieved?

**Comment by philippkuehn** [comment]
Of course there is `nodePasteRule` :)

Add it to `addPasteRules`. You can look at the source files of our core extensions as a starting point.

**Reply by rfgamaral** [reply]
> Of course there is `nodePasteRule` :)

I've searched for that in this repository, and it doesn't find anything ([ref](https://github.com/ueberdosis/tiptap/search?q=nodePasteRule)). There's also no `nodePasteRule.ts` in [this folder](https://github.com/ueberdosis/tiptap/tree/main/packages/core/src/pasteRules).

> Add it to `addPasteRules`. You can look at the source files of our core extensions as a starting point.

I did, and all extensions that uses `addPasteRules`, are **Marks**, not **Nodes**. Am I missing something?

**Reply by philippkuehn** [reply]
ah 🤦‍♂️ thought of `textPasteRule`.

No, there is no `nodePasteRule` because this is way more complex than `nodeInputRule`. Think of multiline markdown blocks like codeblocks or blockquotes. this isn’t parsable that easy. inline nodes are maybe doable.

but you can try to create a plain `new PasteRule` for your specific needs.

**Reply by rfgamaral** [reply]
Thank you once again :)

**Reply by rfgamaral** [reply]
@philippkuehn I'm having a hard time figuring out this one, is there any example on this `new ParseRule` think I could use as base? Either way, let me explain in more detail the use case I have for this, maybe there's a better way to achieve what I want. I actually have two different uses cases:

#### 1. Emoji replacer

I've implemented a custom emoji replacer where if someone types `:D ` (and other emoticons) or if some inputs a unicode emoji (from the system picker), they both get replaced with a custom node to show an image with Twitter Emoji (Twemoji) with the following markup:

```html
<span data-emoji-replacer>
  <img src="https://twemoji.maxcdn.com/v/latest/svg/1f600.svg" alt="😀" class="emoji emoji" draggable="false">
</span>
```

#### 2. Keyword highlighter

I haven't implemented this (but should be easy with `nodeInputRule`), but I need to find specific keywords (based on a complex matching function), for instance "today", and replace them with something like this:

```html
<span data-keyword-highlighter>
  <img src="...">
  <span>today</span>
</span>
```

For both uses cases I need them to work with pasting too, probably with slight different rules (for instance `text:Dtext` shouldn't be replaced, but `text :D text` should). So that if the user pastes text with emojis or specific keywords, those are translated to their respective nodes.

What would be the easiest way to implement such thing? Would really appreciate a basic example on where to start with this. I've been looking deeply into tiptap's extensions source code to understand how to do a lot of the things I want to do, but for this one I'm a bit clueless.

**Reply by philippkuehn** [reply]
Sorry for the confusion. I was on the road earlier and in a hurry. I meant `PasteRule` and not `ParseRule`. PasteRules accepts also custom methods instead of a simple regex. Sometimes a regex is not enough.

It's funny because I've also been working on an official emoji extension for a week now 😅

What I’m doing to detect emojis (typed or pasted):
- add a `Plugin` to `addProseMirrorPlugins`
- hook into `appendTransaction`
- calculate changes ranges
- iterate over ranges to find emojis within text nodes
- replace emoji with custom node

**Reply by rfgamaral** [reply]
> Sorry for the confusion. I was on the road earlier and in a hurry. I meant `PasteRule` and not `ParseRule`.

Ah, no worries, and no need to reply in a hurry, I have time. `PasteRule` makes more sense now, but it will definitely take a bit to implement what I'm looking for (unless you do it first 😅).

> It's funny because I've also been working on an official emoji extension for a week now 😅

That's cool, can't wait to see what you come up with, maybe I'll be able to ditch my own extension or take some ideas of yours :)

**Reply by rfgamaral** [reply]
@philippkuehn I based my paste rules in the `nodeInputRule` and ended up with this:

```ts
addPasteRules() {
    return [
        new PasteRule({
            find: EMOTICON_REGEX_PASTE,
            handler: ({ state, range, match }) => {
                const { tr } = state

                tr.replaceWith(
                    range.from,
                    range.to,
                    this.type.create({
                        emoji: match[0].trim(),
                    }),
                )
            },
        }),
        new PasteRule({
            find: UNICODE_REGEX_PASTE,
            handler: ({ state, range, match }) => {
                const { tr } = state

                tr.replaceWith(
                    range.from,
                    range.to,
                    this.type.create({
                        emoji: match[0].trim(),
                    }),
                )
            },
        }),
    ]
},
```

It seems to be working relatively well for me, see this video:

https://user-images.githubusercontent.com/96476/138154966-a086cc92-4a6a-4d0e-9c27-71ec84ed0f4c.mp4

> No, there is no `nodePasteRule` because this is way more complex than `nodeInputRule`. Think of multiline markdown blocks like codeblocks or blockquotes. this isn’t parsable that easy. inline nodes are maybe doable.

So, what exactly is this problem you mentioned? Can you give me a more complex example where this would fail?

Couldn't we at least have a basic `nodePasteRule` in the core? Maybe with warning stating that it might break for more complex scenarios.
